gapir/cc/stack.h

```diff
@@ -26,457 +26,176 @@
-    }
-    bool getTo(int8_t* i8) const {
-      if (mType != TypeToBaseType<int8_t>::type) {
-        return false;
-      }
-      *i8 = mValue.i8;
-      return true;
-    }
-    bool getTo(int16_t* i16) const {
-      if (mType != TypeToBaseType<int16_t>::type) {
-        return false;
-      }
-      *i16 = mValue.i16;
-      return true;
-    }
-    bool getTo(int32_t* i32) const {
-      if (mType != TypeToBaseType<int32_t>::type) {
-        return false;
-      }
-      *i32 = mValue.i32;
-      return true;
-    }
-    bool getTo(int64_t* i64) const {
-      if (mType != TypeToBaseType<int64_t>::type) {
-        return false;
-      }
-      *i64 = mValue.i64;
-      return true;
-    }
-    bool getTo(uint8_t* u8) const {
-      if (mType != TypeToBaseType<uint8_t>::type) {
-        return false;
-      }
-      *u8 = mValue.u8;
-      return true;
-    }
-    bool getTo(uint16_t* u16) const {
-      if (mType != TypeToBaseType<uint16_t>::type) {
-        return false;
-      }
-      *u16 = mValue.u16;
-      return true;
-    }
-    bool getTo(uint32_t* u32) const {
-      if (mType != TypeToBaseType<uint32_t>::type &&
-          mType != BaseType::VolatilePointer &&
-          mType != BaseType::ConstantPointer) {
-        return false;
-      }
-      *u32 = mValue.u32;
-      return true;
-    }
-    bool getTo(uint64_t* u64) const {
-      if (mType != TypeToBaseType<uint64_t>::type) {
-        return false;
-      }
-      *u64 = mValue.u64;
-      return true;
-    }
-    bool getTo(float* f) const {
-      if (mType != TypeToBaseType<float>::type) {
-        return false;
-      }
-      *f = mValue.f;
-      return true;
-    }
-    bool getTo(double* d) const {
-      if (mType != TypeToBaseType<double>::type) {
-        return false;
-      }
-      *d = mValue.d;
-      return true;
+    template <class T>
+    const T& get() {
+      assertCanStore<T>();
+      assert(sizeof(T) == mDataSize && mDataType == typeid(T));
+      return *((T*)&mData);
     }
 
-    template <typename T>
-    bool getTo(T* val) const {
-      static_assert(sizeof(T) == sizeof(mValue.u32),
-                    "Enum base type is not uint32_t");
-      if (mType != BaseType::Uint32) {
-        return false;
-      }
-      *val = T(mValue.u32);
-      return true;
-    }
+    const std::type_index& getType() const { return mDataType; }
+    bool isVolatilePointer() const { return mIsVolatilePointer; }
 
-    bool getTo(const void** p) const {
-      if (mType != BaseType::AbsolutePointer) {
-        return false;
-      }
-      *p = mValue.p;
-      return true;
-    }
+    void copyTo(void* address) { memcpy(address, (void*)&mData, mDataSize); }
 
-    bool getTo(void** p) const {
-      if (mType != BaseType::AbsolutePointer) {
-        return false;
-      }
-      *p = mValue.p;
-      return true;
+    Entry& operator=(const Entry& rhs) {
+      this->mDataSize = rhs.mDataSize;
+      memcpy((void*)&mData, (void*)&rhs.mData, maxEntrySize);
```

Comment:
I am assuming the only thing that can ever end up in here is a POD type?

Reply:
The code will work with anything that can be shallow copied.